pkg/action/build-iso.go

```diff
@@ -142,23 +136,51 @@ func (b BuildISOAction) prepareISORoot(isoDir string, rootDir string, uefiDir st
 	}
 
 	b.cfg.Logger.Info("Creating EFI image...")
-	err = b.createEFI(uefiDir, filepath.Join(isoDir, constants.IsoEFIPath))
+	err = b.createEFI(rootDir, isoDir)
 	if err != nil {
 		return err
 	}
 	return nil
 }
 
-func (b BuildISOAction) createEFI(root string, img string) error {
-	efiSize, err := utils.DirSize(b.cfg.Fs, root)
+// createEFI creates the EFI image that is used for booting
+// it searches the rootfs for the shim/grub.efi file and copies it into a directory with the proper EFI structure
+// then it generates a grub.cfg that chainloads into the grub.cfg of the livecd (which is the normal livecd grub config from luet packages)
+// then it calculates the size of the EFI image based on the files copied and creates the image
+func (b BuildISOAction) createEFI(rootdir string, isoDir string) error {
+	var err error
+
+	// rootfs /efi dir
+	img := filepath.Join(isoDir, constants.IsoEFIPath)
+	temp, _ := utils.TempDir(b.cfg.Fs, "", "enki-iso")
+	_ = utils.MkdirAll(b.cfg.Fs, filepath.Join(temp, constants.EfiBootPath), constants.DirPerm)
+	_ = utils.MkdirAll(b.cfg.Fs, filepath.Join(isoDir, constants.EfiBootPath), constants.DirPerm)
+
+	err = b.copyShim(temp, rootdir)
 	if err != nil {
 		return err
 	}
 
+	err = b.copyGrub(temp, rootdir)
+	if err != nil {
+		return err
+	}
+
+	// Generate grub cfg that chainloads into the default livecd grub under /boot/grub2/grub.cfg
+	// Its read from the root of the livecd, so we need to copy it into /EFI/BOOT/grub.cfg
+	// This is due to the hybrid bios/efi boot mode of the livecd
+	// the uefi.img is loaded into memory and run, but grub only sees the livecd root
+	_ = b.cfg.Fs.WriteFile(filepath.Join(isoDir, constants.EfiBootPath, constants.GrubCfg), []byte(constants.GrubEfiCfg), constants.FilePerm)
+
+	// Calculate EFI image size based on artifacts
+	efiSize, err := utils.DirSize(b.cfg.Fs, temp)
```

Comment:
maybe not important but I see this function does not check for links, so the final size will be a bit bigger than the actual size

Reply:
No links supported under FAT filesystems, so if there were any links, they would be ignored anyway. Notice that this `temp` dir is what ends up int he efi image, which is formatted with FAT. So there should not be any links here.

Would be nice to have those thougth....would help us with the active/passive/fallback stuff by linking active to the latest and such but no luck there.

Maybe in the future the EFI standard will move on from using FAT fs :(

Reply:
gotcha, thanks for the clarification :) ... yup indeed, and I guess they use such FS to cover any type of OS you may want to boot there?

Reply:
yep, indeed, so both linux,macos and windows can work with EFI stuff. I guess it makes sense as its the most supported Fs ever probably xD